File: js/binaryQuestion/perfectSquare.js
// question : Given a positive integer num, write a function which returns True if num is a perfect square else False.

// O(logN)
function perfectSquare( n ) {
   
    let left = 1;
    let right = n-1;

    while ( left <= right ) {
        const midVal = Math.ceil( ( left + right ) / 2 );
        let squared = midVal * midVal
        if ( squared === n ) return true;

        if ( squared > n ) {
            right = midVal - 1;
        } else {
            left = midVal + 1;
        }
    }

    if ( ( right * right ) === n ) return true
    
    return false


}

console.log(perfectSquare(64))
console.log(perfectSquare(81))
console.log(perfectSquare(4))
console.log(perfectSquare(9))
console.log(perfectSquare(225))
console.log(perfectSquare(144)) 

File: java/binarySearch/perfectSquare.java
package binarySearch;


 // question : Given a positive integer num, write a function which returns True if num is a perfect square else False.
public class perfectSquare {
    

    static boolean perfectSqr(int n) {

        int left = 1;
        int right = n;

        if (left > right)
            return false;

        while (left <= right) {

            int middle = (int) (left + right) / 2;

            if ((middle * middle) == n)
                return true;

            if ((middle * middle) < n) {
                left = middle + 1;
            } else {
                right = middle - 1;
            }
        }

        return false;

    }
    
    public static void main(String[] args) {
        System.out.println(perfectSqr(64));
System.out.println(perfectSqr(81));
System.out.println(perfectSqr(4));
System.out.println(perfectSqr(7));
System.out.println(perfectSqr(235));
System.out.println(perfectSqr(124));
    }
}
